### PR TITLE
Hist-patch2

### PR DESCRIPTION
**Pull request by LovelyBuggies**
look down ⬇️

### CODE REVIEW

**Comment by LovelyBuggies** [comment]
@henryiii It's a little bit later due to my mis-commit. Just focus on the files changed, i.e. Hist.ipynb and src/core.py

**Comment by LovelyBuggies** [comment]
This is just a draft PR and might be closed later, aiming to provide a demo for scipy proceeding slides. The uncertainty plot is not customized and some changes may violate the original testing order. The modification is subjected to more careful tests. 

Btw, some disabled param like bar width in pull_plot is not tested previously. I'll add it in the future.

**Comment by LovelyBuggies** [comment]
Feat:

- [x] remove the mean_plot.
- [ ] add customized unc band.
- [x] allow have empty names, except NamedHist.
- [ ] use `iminuit.minimize`.

**Comment by LovelyBuggies** [comment]
Sometimes, CI throws an error, but sometimes not. Not sure why.

```bash
$ python3 -m pytest
===================================================== test session starts =====================================================
platform darwin -- Python 3.8.3, pytest-5.4.3, py-1.8.1, pluggy-0.13.1
rootdir: /Users/nino/Documents/GitHub/hist
collected 9 items                                                                                                             

tests/test_axis.py ..                                                                                                   [ 22%]
tests/test_basic.py ...                                                                                                 [ 55%]
tests/test_general.py F.                                                                                                [ 77%]
tests/test_named.py ..                                                                                                  [100%]

========================================================== FAILURES ===========================================================
______________________________________________________ test_basic_usage _______________________________________________________

    def test_basic_usage():
        """
            Test basic usage -- whether Hist are properly derived from
            boost-histogram and whether pull_plot method work.
        """
    
        # Basic
        h = Hist(axis.Regular(10, 0, 1, name="x")).fill([0.35, 0.35, 0.45])
    
        for idx in range(10):
            if idx == 3:
                assert h[idx] == h[{0: idx}] == 2
            elif idx == 4:
                assert h[idx] == h[{0: idx}] == 1
            else:
                assert h[idx] == h[{0: idx}] == 0
    
        # Regular
        h = Hist(
            axis.Regular(10, 0, 1, name="x"),
            axis.Regular(10, 0, 1, name="y"),
            axis.Regular(2, 0, 2, name="z"),
        ).fill(
            [0.35, 0.35, 0.35, 0.45, 0.55, 0.55, 0.55],
            [0.35, 0.35, 0.45, 0.45, 0.45, 0.45, 0.45],
            [0, 0, 1, 1, 1, 1, 1],
        )
    
        z_one_only = h[{2: bh.loc(1)}]
        for idx_x in range(0, 10):
            for idx_y in range(0, 10):
                if idx_x == 3 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 1
                elif idx_x == 4 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 1
                elif idx_x == 5 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 3
                else:
                    assert z_one_only[idx_x, idx_y] == 0
    
        # Bool
        h = Hist(axis.Bool(name="x"), axis.Bool(name="y"), axis.Bool(name="z"),).fill(
            [True, True, True, True, True, False, True],
            [False, True, True, False, False, True, False],
            [False, False, True, True, True, True, True],
        )
    
        z_one_only = h[{2: bh.loc(True)}]
        assert z_one_only[False, False] == 0
        assert z_one_only[False, True] == 1
        assert z_one_only[True, False] == 3
        assert z_one_only[True, True] == 1
    
        # Variable
        h = Hist(
            axis.Variable(range(11), name="x"),
            axis.Variable(range(11), name="y"),
            axis.Variable(range(3), name="z"),
        ).fill(
            [3.5, 3.5, 3.5, 4.5, 5.5, 5.5, 5.5],
            [3.5, 3.5, 4.5, 4.5, 4.5, 4.5, 4.5],
            [0, 0, 1, 1, 1, 1, 1],
        )
    
        z_one_only = h[{2: bh.loc(1)}]
        for idx_x in range(0, 10):
            for idx_y in range(0, 10):
                if idx_x == 3 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 1
                elif idx_x == 4 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 1
                elif idx_x == 5 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 3
                else:
                    assert z_one_only[idx_x, idx_y] == 0
    
        # Integer
        h = Hist(
            axis.Integer(0, 10, name="x"),
            axis.Integer(0, 10, name="y"),
            axis.Integer(0, 2, name="z"),
        ).fill(
            [3.5, 3.5, 3.5, 4.5, 5.5, 5.5, 5.5],
            [3.5, 3.5, 4.5, 4.5, 4.5, 4.5, 4.5],
            [0, 0, 1, 1, 1, 1, 1],
        )
    
        z_one_only = h[{2: bh.loc(1)}]
        for idx_x in range(0, 10):
            for idx_y in range(0, 10):
                if idx_x == 3 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 1
                elif idx_x == 4 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 1
                elif idx_x == 5 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 3
                else:
                    assert z_one_only[idx_x, idx_y] == 0
    
        # IntCategory
        h = Hist(
            axis.IntCategory(range(10), name="x"),
            axis.IntCategory(range(10), name="y"),
            axis.IntCategory(range(2), name="z"),
        ).fill(
            [3.5, 3.5, 3.5, 4.5, 5.5, 5.5, 5.5],
            [3.5, 3.5, 4.5, 4.5, 4.5, 4.5, 4.5],
            [0, 0, 1, 1, 1, 1, 1],
        )
    
        z_one_only = h[{2: bh.loc(1)}]
        for idx_x in range(0, 10):
            for idx_y in range(0, 10):
                if idx_x == 3 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 1
                elif idx_x == 4 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 1
                elif idx_x == 5 and idx_y == 4:
                    assert z_one_only[idx_x, idx_y] == 3
                else:
                    assert z_one_only[idx_x, idx_y] == 0
    
        # StrCategory
        h = Hist(
            axis.StrCategory("FT", name="x"),
            axis.StrCategory(list("FT"), name="y"),
            axis.StrCategory(["F", "T"], name="z"),
        ).fill(
            ["T", "T", "T", "T", "T", "F", "T"],
            ["F", "T", "T", "F", "F", "T", "F"],
            ["F", "F", "T", "T", "T", "T", "T"],
        )
    
        z_one_only = h[{2: bh.loc("T")}]
        assert z_one_only[bh.loc("F"), bh.loc("F")] == 0
        assert z_one_only[bh.loc("F"), bh.loc("T")] == 1
        assert z_one_only[bh.loc("T"), bh.loc("F")] == 3
        assert z_one_only[bh.loc("T"), bh.loc("T")] == 1
    
        # right pull_plot
        h = Hist(
            axis.Regular(
                50, -4, 4, name="S", title="s [units]", underflow=False, overflow=False
            )
        ).fill(np.random.normal(size=1_000))
    
        def pdf(x, a=1 / np.sqrt(2 * np.pi), x0=0, sigma=1, offset=0):
            exp = unp.exp if a.dtype == np.dtype("O") else np.exp
            return a * exp(-((x - x0) ** 2) / (2 * sigma ** 2)) + offset
    
>       assert h.pull_plot(
            pdf,
            eb_ecolor="crimson",
            eb_mfc="crimson",
            eb_mec="crimson",
            eb_fmt="o",
            eb_ms=6,
            eb_capsize=1,
            eb_capthick=2,
            eb_alpha=0.8,
            vp_c="gold",
            vp_ls="-",
            vp_lw=8,
            vp_alpha=0.6,
            fp_c="chocolate",
            fp_ls="-",
            fp_lw=3,
            fp_alpha=1.0,
            bar_fc="orange",
            pp_num=6,
            pp_fc="orange",
            pp_alpha=0.618,
            pp_ec=None,
        )

tests/test_general.py:158: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
src/hist/core.py:59: in pull_plot
    copt = correlated_values(popt, pcov)
../../../anaconda3/envs/hist/lib/python3.8/site-packages/uncertainties/core.py:181: in correlated_values
    return correlated_values_norm(
../../../anaconda3/envs/hist/lib/python3.8/site-packages/uncertainties/core.py:230: in correlated_values_norm
    (variances, transform) = numpy.linalg.eigh(correlation_mat)
<__array_function__ internals>:5: in eigh
    ???
../../../anaconda3/envs/hist/lib/python3.8/site-packages/numpy/linalg/linalg.py:1446: in eigh
    w, vt = gufunc(a, signature=signature, extobj=extobj)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

err = 'invalid value', flag = 8

    def _raise_linalgerror_eigenvalues_nonconvergence(err, flag):
>       raise LinAlgError("Eigenvalues did not converge")
E       numpy.linalg.LinAlgError: Eigenvalues did not converge

../../../anaconda3/envs/hist/lib/python3.8/site-packages/numpy/linalg/linalg.py:103: LinAlgError
====================================================== warnings summary =======================================================
/Users/nino/anaconda3/envs/hist/lib/python3.8/site-packages/past/builtins/misc.py:45
  /Users/nino/anaconda3/envs/hist/lib/python3.8/site-packages/past/builtins/misc.py:45: DeprecationWarning: the imp module is deprecated in favour of importlib; see the module's documentation for alternative uses
    from imp import reload

tests/test_basic.py::test_basic_usage
tests/test_basic.py::test_errors
tests/test_general.py::test_errors
tests/test_named.py::test_basic_usage
tests/test_named.py::test_errors
  /Users/nino/Documents/GitHub/hist/src/hist/core.py:65: RuntimeWarning: divide by zero encountered in true_divide
    pulls = (self.view() - y_nv) / yerr

tests/test_general.py::test_basic_usage
  /Users/nino/anaconda3/envs/hist/lib/python3.8/site-packages/scipy/optimize/minpack.py:807: OptimizeWarning: Covariance of the parameters could not be estimated
    warnings.warn('Covariance of the parameters could not be estimated',

tests/test_general.py::test_basic_usage
  /Users/nino/anaconda3/envs/hist/lib/python3.8/site-packages/uncertainties/core.py:185: RuntimeWarning: invalid value encountered in true_divide
    covariance_mat/norm_vector/norm_vector[:,numpy.newaxis],

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=================================================== short test summary info ===================================================
FAILED tests/test_general.py::test_basic_usage - numpy.linalg.LinAlgError: Eigenvalues did not converge
=========================================== 1 failed, 8 passed, 8 warnings in 1.49s ===========================================
```

File: src/hist/core.py [hunk 1]
```diff
@@ -13,12 +14,11 @@ def __init__(self, *args, **kwargs):
             Initialize Hist object. Axis params must contain the names.
         """
 
+        # ToDo: why cannot use super(Histogram, self).__init__(*args, **kwargs)
         super(BaseHist, self).__init__(*args, **kwargs)
```
Comment (henryiii) [inline src/hist/core.py:18]: Can't you use this:

```suggestion
        super().__init__(*args, **kwargs)
```

Comment (LovelyBuggies) [reply src/hist/core.py:18]: We can use this. But I don't know why this happens.

**Pull request merged**